### PR TITLE
feat(aws): New Neptune, ElastiCache, APIGW and IAM checks

### PR DESCRIPTION
**Pull request by jit-contrib**
### Context

Adding support for some new rules for AWS provider.

### Description

This PR adds the following new AWS rules:

* neptune_uses_a_public_subnet
* elasticache_using_public_subnets
* ecs_task_definitions_public_ip -- **This check has been deleted from the PR since using the task definition you cannot verify if the container/task definition has a public IP. Also the `network_mode` `awsvpc` is mandatory if you are using FARGATE.**
* apigateway_endpoint_public_without_authorizer
* iam_use_temporary_credentials

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by MrCloudSec** [comment]
Great job @jit-contrib ! Could you fill the metadata values that are empty pls?
Also, can you add a filter in the Neptune service to only list the clusters with a Neptune engine? Thanks!

**Comment by MrCloudSec** [comment]
Also, can you change the check names so they follow the pattern of `service_subservice`?
For example, `neptune_uses_a_public_subnet` -> `neptune_cluster_with_public_subnet`.

File: prowler/providers/aws/services/elasticache/elasticache_using_public_subnets/elasticache_using_public_subnets.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        findings = []
+        vpc_subnets = vpc_client.vpc_subnets
+        for instance in elasticache_client.elasticache_instances:
+            report = Check_Report_AWS(self.metadata())
+            report.resource_id = instance.cache_cluster_id
+            report.resource_arn = instance.arn
+            report.status = "PASS"
+            report.status_extended = "Cluster isn't using public subnets."
+            public_subnets = []
+            # print(instance.subnet_group)
```
Comment (jfagoagas) [inline prowler/providers/aws/services/elasticache/elasticache_using_public_subnets/elasticache_using_public_subnets.py]:
```suggestion
```

File: prowler/providers/aws/services/elasticache/elasticache_using_public_subnets/elasticache_using_public_subnets.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+            for subnets in instance.subnet_group:
+                for subnet in subnets["Subnets"]:
+                    if vpc_subnets[subnet["SubnetIdentifier"]].public:
+                        public_subnets.append(
+                            vpc_subnets[subnet["SubnetIdentifier"]].id
+                        )
```
Comment (jfagoagas) [inline prowler/providers/aws/services/elasticache/elasticache_using_public_subnets/elasticache_using_public_subnets.py]: This logic should be moved to the Elasticache service.

File: prowler/providers/aws/services/iam/iam_service.py [hunk 3]
```diff
@@ -661,6 +665,27 @@ def __list_tags_for_resource__(self):
                 f"{self.region} -- {error.__class__.__name__}[{error.__traceback__.tb_lineno}]: {error}"
             )
 
+    def __get_last_accessed_services__(self):
+        for user in self.users:
+            details = self.client.generate_service_last_accessed_details(Arn=user.arn)
+            response = self.client.get_service_last_accessed_details(
+                JobId=details["JobId"]
+            )
+            while response["JobStatus"] == "IN_PROGRESS":
```
Comment (jfagoagas) [inline prowler/providers/aws/services/iam/iam_service.py]: Why do we need to do the `while` block?

File: prowler/providers/aws/services/iam/iam_service.py [hunk 4]
```diff
@@ -661,6 +665,27 @@ def __list_tags_for_resource__(self):
                 f"{self.region} -- {error.__class__.__name__}[{error.__traceback__.tb_lineno}]: {error}"
             )
 
+    def __get_last_accessed_services__(self):
+        for user in self.users:
```
Comment (jfagoagas) [inline prowler/providers/aws/services/iam/iam_service.py]: Please include two `try/except` blocks here, one above `for user in self.users:` and another below not to stop the iteration if something fails.

File: prowler/providers/aws/services/iam/iam_service.py [hunk 5]
```diff
@@ -661,6 +665,27 @@ def __list_tags_for_resource__(self):
...
+            )
+            while response["JobStatus"] == "IN_PROGRESS":
+                response = self.client.get_service_last_accessed_details(
+                    JobId=details["JobId"]
+                )
+            self.last_accessed_services.append(
+                {"user": user, "services": response["ServicesLastAccessed"]}
+            )
+
+    def __get_access_keys_metadata__(self):
```
Comment (jfagoagas) [inline prowler/providers/aws/services/iam/iam_service.py:708]: Please include two `try/except` blocks here, one above `for user in self.users:` and another below not to stop the iteration if something fails.

File: prowler/providers/aws/services/iam/iam_use_temporary_credentials/iam_use_temporary_credentials.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+                    [
+                        akm
+                        for akm in iam_client.access_keys_metadata[idx][
+                            "access_keys_metadata"
+                        ]
+                        if len(akm["AccessKeyMetadata"]) > 0
+                    ]
+                )
+                > 0
+            ):
```
Comment (jfagoagas) [inline prowler/providers/aws/services/iam/iam_use_temporary_credentials/iam_use_temporary_credentials.py]: Could we move this block to the IAM service?

File: prowler/providers/aws/services/neptune/neptune_service.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    def __init__(self, audit_info):
+        # Call AWSService's __init__
+        super().__init__(__class__.__name__, audit_info)
+        self.clusters = []
+        self.__describe_clusters__()
+
+    def __describe_clusters__(self):
+        logger.info("Neptune - Describing DB Clusters...")
+        try:
+            for cluster in self.client.describe_db_clusters()["DBClusters"]:
```
Comment (jfagoagas) [inline prowler/providers/aws/services/neptune/neptune_service.py]: Do we need to filter the clusters just for the neptune ones.

File: prowler/providers/aws/services/neptune/neptune_service.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+            logger.error(
+                f"{error.__class__.__name__}[{error.__traceback__.tb_lineno}]: {error}"
+            )
+
+
+class Cluster(BaseModel):
+    arn: str
+    name: str
+    id: str
+    subnet_group: typing.Any
```
Comment (jfagoagas) [inline prowler/providers/aws/services/neptune/neptune_service.py]: Why this is `typing.Any`?

File: prowler/providers/aws/services/neptune/neptune_uses_a_public_subnet/neptune_uses_a_public_subnet.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+            report.resource_arn = cluster.arn
+            report.status = "PASS"
+            report.status_extended = "Cluster isn't using public subnets."
+            public_subnets = []
+            for subnets in cluster.subnet_group:
+                for subnet in subnets["Subnets"]:
+                    if vpc_subnets[subnet["SubnetIdentifier"]].public:
+                        public_subnets.append(
+                            vpc_subnets[subnet["SubnetIdentifier"]].id
+                        )
```
Comment (jfagoagas) [inline prowler/providers/aws/services/neptune/neptune_uses_a_public_subnet/neptune_uses_a_public_subnet.py]: This should be move to the Neptune service.

**Review by jfagoagas — CHANGES_REQUESTED** [review]
I left some comments to complement Sergio's review.

File: prowler/providers/aws/services/iam/iam_service.py (hunk 3 above)
Comment (jit-contrib) [reply prowler/providers/aws/services/iam/iam_service.py]: API returns while the job to generate_service_last_accessed_details is ongoing with the JobStatus "IN_PROGRESS", there is a need to call the API until the JobStatus changed from "IN_PROGRESS" to "COMPLETED"

**Review by jfagoagas — APPROVED** [review]
🚀 🚀🚀

**Pull request merged**